File: Leetcode Medium/200NumberOfIslands.py
#https://leetcode.com/problems/number-of-islands/description/
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def dfs(i, j):
            if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or grid[i][j] == "0":
                return
            else:
                grid[i][j] = "0"
                dfs(i+1, j)
                dfs(i-1, j)
                dfs(i, j+1)
                dfs(i, j-1)
        counter = 0
        for i, row in enumerate(grid):
            for j, land in enumerate(row):
                if land == "1":
                    counter += 1
                    dfs(i, j)
                    grid[i][j] = "0"
        return counter

File: C++/Leetcode Easy/200NumberOfIslands.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int res = 0;
        for (int i = 0; i < grid.size(); ++i){
            for (int j = 0; j < grid[0].size(); ++j){
                if (grid[i][j] == '0') continue;
                ++res;
                helper(i, j, grid);
            }
        }
        return res;
    }
    void helper (int i, int j, vector<vector<char>>& grid){
        if (i < 0 || i >= grid.size()) return;
        if (j < 0 || j >= grid[0].size()) return;
        if (grid[i][j] == '0') return;
        grid[i][j] = '0';
        helper (i-1, j, grid);
        helper (i+1, j, grid);
        helper (i, j-1, grid);
        helper (i, j+1, grid);
    }
};